### DISCUSSION TITLE
Migrate to Active Guidance Info from NavNative?

### DISCUSSION THREAD

**Discussion by S2Ler** [mapbox/mapbox-navigation-ios#3476] [Q&A]
The Navigation Status contains `activeGuidanceInfo` property[^1]

It contains information route progress (https://github.com/mapbox/mapbox-navigation-native/blob/master/bindings/generated/objc/MBNNActiveGuidanceProgress.h) for route/leg/step, like:
- distanceTraveled
- fractionTraveled
- remainingDistance
- remainingDuration

We don't use this information and calculate it on the SDK side. I prototyped possible implementation[^2] and tested it in the Example app. I haven't found apparent issues, although tests were failing (RouteProgressTests and this is expected).

The question is shall we migrate to MBNNActiveGuidanceInfo?

@mapbox/navigation-ios 

[^1]: https://github.com/mapbox/mapbox-navigation-native/blob/506ca792c4ff2bd98c0f2b2b0f59d98fb2617844/bindings/generated/objc/MBNNNavigationStatus.h#L135
[^2]: https://github.com/mapbox/mapbox-navigation-ios/tree/feature/active-guidance-info

**Comment by 1ec5** [comment]
These four aggregate properties alone don’t make for a particularly compelling case to migrate wholesale to MBNNActiveGuidanceInfo, since they’re nothing more than basic arithmetic. In fact, in order to avoid wrapping the C++ object, we’d end up introducing quite a bit of parallel state that can easily get out of sync. But there may be other properties worth copying into the progress classes on each location update, as we currently do with some of the RouteStepProgress properties.

**Reply by S2Ler** [reply]
But we duplicate work, don't we? Many properties are calculated on each access and for long routes it can produce a lot of overhead.

**Reply by 1ec5** [reply]
In principle yes, but we’d want to profile the impact to understand the tradeoff better, because there is a real risk associated with introducing parallel state. The route progress model objects were patterned on the Progress class in Foundation, which aggregates values in the same manner. Someday it would be nice to be able to migrate closer to that model rather than away from it.

**Reply by S2Ler** [reply]
I mean, we keep our RouteProgress, but instead of calculating values manually (hello LegacyRouteController), we pick the data from NavNative that should be the source of truth anyway.

**Reply by truburt** [reply]
As discussed today during the NN-NSDK sync, Android has been utilizing the progress data from NN for a while now with no issues. We should build PoC on iOS and see if there're any risks or problems along the way.
@S2Ler , could you please cut a ticket for that? We'll then be good to close this discussion.

**Comment by jahangir520** [comment]
ab kuch nai ho sakta
